### DISCUSSION TITLE
Picosnitch on immutable systems

### DISCUSSION THREAD

**Discussion by quazar-omega** [elesiuta/picosnitch#29] [General]
Hi! Sorry for being so late to actually respond to your message, I was quite busy.  
In my environment (Fedora toolbox) I tried the built-in daemon and while the database is created it is never populated, I also can't see the daemon running after I launch it.

About possibly contributing, I'm fairly a noob in Python so I was trying to understand how the project works, but I wasn't able to set it up, is it possible to use poetry+pyenv with this somehow? Or do I have to use the system installed python directly?

**Comment by elesiuta** [comment]
No worries, same here. 

The daemon is probably crashing because it is unable to load the BPF program which does the system monitoring, and therefore no entries to populate the database. Here are some small BPF programs you can use to test whether you have BCC installed properly and is accessible from your environment [https://github.com/iovisor/bcc/tree/master/tools](https://github.com/iovisor/bcc/tree/master/tools), simply run any of the .py files as root (they might also already be installed and runnable with `sudo bcc.tool_name`). 

I welcome all contributions and am happy to help. The main dependency is BCC which needs to be installed as a system package since it can't be done through PyPI. The only other required dependency is `psutil` which can be installed via your system python or as a user package. Due to BCC, picosnitch required using your system installed python directly, but I just pushed some commits today which will allow picosnitch find your system's BCC when picosnitch is installed in a venv or with pipx (I haven't tested, nor use poetry+pyenv but see no reason why it shouldn't work after today's changes), you can also skip the install and just run the file picosnitch.py.

To help with navigating the source code, I have a [callgraph here](https://raw.githubusercontent.com/elesiuta/picosnitch/master/docs/callgraph.svg), in summary the program command line starts with the function `start_picosnitch` which then calls `ui_dash`, `ui_init`, or `main` for the web dashboard, terminal interface, or daemon respectively. `main` calls `main_process` which starts up and monitors 5 sub-processes, each their own function, which run for the life of the program.
